src/cart/crt_hg.c

```diff
@@ -623,13 +628,54 @@ d_list_t
 	return &prov_data->cpg_ctx_list;
 }
 
+static int
+crt_get_opx_info_string(char *provider, char *domain, char *ip,
+			char **string, int start_port, int ctx_idx)
+{
+	int	rc = 0;
+	int	hfi = -1;
+	int	delimiter;
+	char	*hfi_str = NULL;
+	char	domain_name[10];
+
+	/* Current support for the following domains: ib<hfi> or opx<hfi> */
+	if (strncmp(domain, "ib", 2) == 0)
+		delimiter = 1;
+	else if (strncmp(domain, "opx", 3) == 0)
+		delimiter = 2;
+	else {
+		D_ERROR("Invalid OPX domain name.\n");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	strcpy(domain_name, domain);
+	strtok_r(domain_name, &domain[delimiter], &hfi_str);
+	hfi = (unsigned int)strtoul(hfi_str, NULL, 10);
+
+	if (ip == NULL)
+		D_ASPRINTF(*string, "%s://%s:%d:%d",
+			   provider, domain, hfi,
+			   start_port + ctx_idx);
+	else
+		D_ASPRINTF(*string, "%s://%s/%s:%d:%d",
+			   provider, domain, ip, hfi,
+			   start_port + ctx_idx);
+
+out:
+	if (!rc && *string == NULL)
+		return -DER_NOMEM;
```

Comment:
Seems from this as if we should already be returning -DER_NOMEM for all out-of-memory cases. Not sure how we could get into the case that I commented on below but perhaps other providers are?